### PR TITLE
[Feature] Fallback controllers

### LINKED ISSUE #1468: Fallback controllers
1. PR: Add information about controller chain into `ControllerInfo` structure
2. PR: Add new parameter under controller that replaces that controller with a list of controllers
3. PR: Implement the feature

### PR DESCRIPTION
**Pull request by saikishor**
We already introduced the necessary information for setting up the fallback controllers via (#1499 and #1503). This PR aims to complete the Fallback controllers feature (When a controller is returning ERROR in the update cycle, it will be deactivated and any fallback controllers linked to this controller will be activated)

This PR has the main logic to have the fallback controllers working, pending some tests
 
Closes #1468

### CODE REVIEW

File: controller_manager/include/controller_manager/controller_manager.hpp [hunk 1]
```diff
@@ -418,6 +424,18 @@ class ControllerManager : public rclcpp::Node
     const std::vector<ControllerSpec> & controllers, int strictness,
     const ControllersListIterator controller_it);
 
+  /// The methood to check if the fallback controllers of the given controllers are in the right
```
Comment (bmagyar) [inline controller_manager/include/controller_manager/controller_manager.hpp:427]:
```suggestion
  /// Check if the fallback controllers of the given controllers are in the right
```

File: controller_manager/src/controller_manager.cpp [hunk 2]
```diff
@@ -2360,16 +2409,67 @@ controller_interface::return_type ControllerManager::update(
   }
   if (!failed_controllers_list.empty())
   {
-    std::string failed_controllers;
+    std::vector<std::string> active_controllers_using_interfaces(failed_controllers_list);
+    active_controllers_using_interfaces.reserve(500);
+    std::vector<std::string> cumulative_fallback_controllers;
+    cumulative_fallback_controllers.reserve(500);
```
Comment (bmagyar) [inline controller_manager/src/controller_manager.cpp:2415]:
```suggestion
    const auto FALLBACK_STACK_MAX_SIZE = 500;
    std::vector<std::string> active_controllers_using_interfaces(failed_controllers_list);
    active_controllers_using_interfaces.reserve(FALLBACK_STACK_MAX_SIZE);
    std::vector<std::string> cumulative_fallback_controllers;
    cumulative_fallback_controllers.reserve(FALLBACK_STACK_MAX_SIZE);
```

File: controller_manager/src/controller_manager.cpp [hunk 3]
```diff
@@ -2360,16 +2409,67 @@ controller_interface::return_type ControllerManager::update(
...
+        for (const auto & fallback_controller : ctrl_it->info.fallback_controllers_names)
+        {
+          cumulative_fallback_controllers.push_back(fallback_controller);
+          get_active_controllers_using_command_interfaces_of_controller(
+            fallback_controller, rt_controller_list, active_controllers_using_interfaces);
+        }
+      }
+    }
+    std::string controllers_string;
+    controllers_string.reserve(500);
```
Comment (bmagyar) [inline controller_manager/src/controller_manager.cpp:2433]:
```suggestion
    controllers_string.reserve(FALLBACK_STACK_MAX_SIZE);
```

Comment (bmagyar) [inline controller_manager/src/controller_manager.cpp:2433]: quite interesting that there isn't a pre-reserve constructor for strings...

**Review by bmagyar — DISMISSED** [review]

File: controller_manager/src/controller_manager.cpp (hunk 3 above)
Comment (saikishor) [reply controller_manager/src/controller_manager.cpp:2433]: Yes it sucks. I want to optimize all of these sooner in the following PRs

Comment (destogl) [reply controller_manager/src/controller_manager.cpp:2434]: Can't we pre-reserve all of this?

**Review by destogl — APPROVED** [review]
Looks good. We should proceed and then optimize the vector and string initialization part.

**Comment by destogl** [comment]
Merging this as discussed with @bmagyar. We should then optimize the memory allocation

**Pull request merged**